[Irfan Danial] iP

idgrr opened the pull request:
DukePro

>“Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,

* text-based
* easy to learn
* ~~FAST~~ _**SUPER FAST**_ to use

All you need to do is,

1. download it from here.
1. double-click it.
1. add your tasks.
1.  let it manage your tasks for you 😉

And it is **FREE!**

Features:

- [x] Managing tasks
- [ ]  Managing deadlines (coming soon)
- [ ]  Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the main method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

idgrr committed:
Added basic response functionality

idgrr committed:
Updated level-1 using oop principles

idgrr committed:
Added Add/ list functionality

idgrr committed:
Changed storage data structure ( array to linked hashmap, Task class)

idgrr committed:
Added Done feature

idgrr committed:
Added Todo, Event deadline functionality

idgrr committed:
Cleaned up the program as well as added JavaDocs

idgrr committed:
added a simple output format option

idgrr committed:
Cleaned up the speech bubble

idgrr committed:
Completed Ui testing

idgrr committed:
Added custom exceptions

idgrr committed:
Added Delete Functionality

idgrr committed:
added JUnit5 files

idgrr committed:
Updated minor styling

idgrr committed:
Added Saving functionality

Current State :
Previously did not save output messages

Problem with current state:
SInce did not save any of session's code, could not access history of current session

How commit changes:
Added a writer function to keep track of the CURRENT session's output messages.
Previous sessions's output could be obtained BEFORE initialising WRITER ( due to Writer resetting file )

idgrr committed:
Updated Task oop

Currently state:
Task's subclasses seperately have common variables such as time and eventtype

Problems with current state:
Seperatting common variables makes it hard to change formating for task in genera;

Commit changes:
Have common variables such as time and eventtype to be stored in task class

idgrr committed:
Add date and time functionality

Previously captured date inputted as a string. Was not able to
use date to compare tasks.

This commit allows time (formatted in YYYY MM DD) to be stored as
a LocalDate so that two task can be compared as to which task
is more urgent

idgrr committed:
Merge branch 'master' into branch-Level-7

idgrr committed:
Merge branch 'branch-Level-7'

idgrr committed:
Merge branch 'master' into branch-Level-8

idgrr committed:
Organize program to incorporate more OOP

idgrr committed:
Fix Storage functionality

Previously did not automatically restore the past messages

idgrr committed:
Package classes into various components

idgrr committed:
Fix output message

idgrr committed:
Merge branch 'master' into branch-Level-7

idgrr committed:
Fix branching problems

idgrr committed:
Merge branch 'branch-Level-7'

idgrr committed:
Fix time display format

Currently does not fultil requirement of of changed date format

Commit includes getFormatDate() to get a formatted version of input
date

idgrr committed:
Fix filepath definition to be outside of storage class

idgrr committed:
Add more exceptions

Currently exception name does not provide useful information

Added exceptions that describes the error occurred better

idgrr committed:
Add JUnit test for Deadline.java

Currently no JUnit test for Deadline

JUnit test added test for exceptions and method outputs

idgrr committed:
Create jar file for project

idgrr committed:
Adds JavaDoc for classes under the util package

idgrr committed:
Adds JavaDocs for classes in tasktypes package

idgrr committed:
Update all the javadocs

idgrr committed:
Fix minor coding standard violations

idgrr committed:
Merge branch 'branch-A-JavaDoc'

idgrr committed:
Merge branch 'master' into branch-A-CodingStandard

# Conflicts:
#	src/main/java/Duke.java
#	src/main/java/duke/command/AddCommand.java
#	src/main/java/duke/command/ByeCommand.java
#	src/main/java/duke/command/ListCommand.java
#	src/main/java/duke/exception/DukeException.java
#	src/main/java/duke/taskTypes/Deadline.java
#	src/main/java/duke/taskTypes/Event.java
#	src/main/java/duke/taskTypes/Task.java

idgrr committed:
Merge branch 'branch-A-CodingStandard'

idgrr committed:
Add find functionality

idgrr committed:
Merge branch 'master' into branch-Level-9

# Conflicts:
#	src/main/java/duke/util/Parser.java
#	src/main/java/duke/util/TaskList.java

idgrr committed:
Merge branch 'branch-Level-9'

yongchuann commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,43 @@
+ * Command that contains details when deleting a task
+ */
+public class DeleteCommand extends Command {
+    private final String chosenList;
```
Is it weird to call this chosenList when you are deleting a task? :flushed:
Maybe you can call it chosenTask or taskNumber?
Same for the doneCommand as well!

idgrr replied:
Yeah misnaming on my part. Thanks for pointing it out!

yongchuann commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    /**
+     * Executes a set of instructions
```
Maybe you can say what exactly the instructions are
Same goes for the other commands!

yongchuann commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,59 @@
+
+    @Override
+    public boolean exec() throws DukeException {
+        LinkedHashMap mapper = taskList.getMapper();
```
Good idea using HashMap to find!

yongchuann commented on src/main/java/duke/exception/EmptyTimeException.java:
```diff
@@ -0,0 +1,16 @@
+/**
+ * Exception that is thrown when user inputs missing time
+ */
+public class EmptyTimeException extends DukeException{
```
I like how you created exceptions for possible user errors

yongchuann commented on src/main/java/duke/taskTypes/Deadline.java:
```diff
@@ -0,0 +1,62 @@
+     */
+    public Deadline(String input, boolean isDone) throws DukeException {
+        super(isDone);
+        List<String> results = Pattern.compile("/by").splitAsStream(input).map(x->x.trim()).collect(Collectors.toList());
```
I don't know if its GitHub or your code but the indentation for this seems a bit off, maybe you can consider splitting it at each new function so that its easier to read?
E.g:
~~~
    Pattern.compile("/by")
            .splitAsStream(input)
~~~

yongchuann commented on src/main/java/duke/util/TaskList.java:
```diff
@@ -0,0 +1,205 @@
+        try {
+            int list_no = Integer.parseInt(input.trim()); //possible NumberFormatException
+            Collection<Task> values = mapper.values();
+            Task second = values.stream().skip(list_no-1).findFirst().orElse(null); // possible for task to be null
```
Good job catching the null here!

yongchuann commented on src/main/java/duke/util/Storage.java:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    /**
+     * Updates and saves the state of the changed task
```
I think you are missing an empty line between the comment and the @param here

yongchuann commented on src/main/java/duke/command/ByeCommand.java:
```diff
@@ -0,0 +1,29 @@
+    @Override
+    public boolean exec() throws DukeException {
+        ui.goodbye();
+        return false;
```
Can I ask why did you choose to make the execute function return the boolean instead of setting it as a parameter?

idgrr replied:
Yup, I agree. Didn't think about it at that time. Thanks for the suggestion!

yongchuann reviewed (commented):
Looks good to me! Just some small things that I've mentioned.
Do also update the description.
If you disagree do let me know as well as i want to learn as well!

idgrr committed:
Merge branch 'A-Gradle' into add-gradle-support

idgrr committed:
Adds Gradle Dependencies

idgrr committed:
Merge branch 'add-gradle-support' into A-Gradle

idgrr committed:
Fix minor exceptions

idgrr committed:
Add Gradle files

idgrr committed:
Get latest from Master

idgrr committed:
Merge branch 'A-Gradle'

# Conflicts:
#	gradlew

idgrr committed:
Remove UI class to keep duke output simple

Previously duke's output had a lot of additional formatting to
incorporate "personality"

Using javaFX, a formatted output seems counter productive and
a simpler output would allow more styling flexibility.

This commit will return duke's response without additional
formatting

okyntary commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,64 @@
+import duke.command.Command;
+
+import duke.exception.DukeException;
+
+import duke.util.Parser;
+import duke.util.Storage;
+import duke.util.TaskList;
+import duke.util.Ui;
```
I never realized my classes could be organised this way (though in hindsight, it does seem kind of obvious). Thanks for the inspiration.

okyntary commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,64 @@
+     */
+    public Duke(String filePath) throws DukeException{
+        storage = new Storage(filePath);
+        ui = new Ui(48);
```
Is there a reason for the ```Ui()``` constructor to need an argument of ```48```?

idgrr replied:
Duke's output was wrapped around this speech bubble format. 48 was a default spacing size for the speech bubble. Any longer than 48 would require reformating and extension of the the speech bubble width. Should have made it into a static variable in UI itself to avoid confusion/add an explanation. Thanks for pointing it out!

okyntary commented on src/main/java/duke/command/ByeCommand.java:
```diff
@@ -0,0 +1,29 @@
+     * @param taskList Tasklist to add task to
+     * @param ui Ui to display msg
+     */
+    public ByeCommand(Storage storage, TaskList taskList,Ui ui){
```
Very small issue: there should be a space before the ```Ui```.

okyntary commented on src/main/java/duke/command/DoneCommand.java:
```diff
@@ -0,0 +1,42 @@
+    /**
+     * Executes a set of instructions
+     *
+     * @return boolean To relay whether to continue the project
```
I'm a little confused about what "continue the project" means here, could you clarify?

idgrr replied:
The returned boolean would let duke decide whether to continue the while loop in Duke.java .and continue waiting for user input. Thinking back, this boolean could be set as a variable in Task and checking could be done by a method call.

okyntary reviewed (commented):
It looks pretty good. It's quite long/wordy, though; I'll admit I didn't read through everything.

idgrr committed:
Add gui functionality

idgrr committed:
Update GUI

idgrr committed:
Merge branch 'branch-A-Level-10'

idgrr committed:
Minor Styling changes

idgrr committed:
Fix Launcher class

There were previously some issues with duke being unable to run

Created a launcher class to set as a starting point for application

This is so that there is a clear starting point for application to
start

idgrr committed:
Add Assertions into taskTypes

Previously there were assumptions on the length of the formatted
input. There were

Adding assumptions would remind collaborators that the expected
formatted input size

This is to ensure that there will not be any unforeseen exceptions
caused by incorrect formatted input.

idgrr committed:
Merge pull request #1 from idgrr/branch-A-Assertions

Add Assertions into taskTypes

idgrr committed:
Refactor command files

Previously was not refactored. Could cause confusion during code
review

Refactoring and splitting some methods would make things clearer

With refactoring, would mean less time understanding code and more
time to review code

idgrr committed:
Refactor exception files

Previously was not refactored. Could cause confusion during code
review

Refactoring and splitting some methods would make things clearer

With refactoring, would mean less time understanding code and more
time to review code

idgrr committed:
Refactor gui files

Previously was not refactored. Could cause confusion during code
review

Refactoring and splitting some methods would make things clearer

With refactoring, would mean less time understanding code and more
time to review code

idgrr committed:
Refactor tasktypes files

Previously was not refactored. Could cause confusion during code
review

Refactoring and splitting some methods would make things clearer

With refactoring, would mean less time understanding code and more
time to review code

idgrr committed:
Merge pull request #2 from idgrr/branch-A-CodeQuality

Branch a code quality

idgrr committed:
Allow setting csv as a source file

idgrr committed:
Merge branch 'branch-BCD-Extension'

idgrr committed:
Fix saving functionality

Previously there were some issues with saving into both csv and txt
files, especially when closing straightaway

Commit fixes this by formatting task save functionality as well as
including application closing event

idgrr committed:
Set theme jekyll-theme-slate

idgrr committed:
Cleaned up the code to improve code quality

idgrr committed:
Merge branch 'master' of https://github.com/idgrr/ip

idgrr committed:
Fix markdown formatting in README.md

idgrr committed:
Fix more Markdown formatting

idgrr committed:
Fix more Markdown Formatting

idgrr committed:
Fix Markdown formatting for images

idgrr committed:
Fix Misaligned pictures

idgrr committed:
Updated UI picture

idgrr committed:
Fix GUI formatting

idgrr committed:
Removed MarkDown images

idgrr committed:
Added Markdown images

idgrr committed:
Update README.md

idgrr committed:
Edit Markdown format

idgrr committed:
Merge branch 'master' of https://github.com/idgrr/ip

idgrr committed:
Set theme jekyll-theme-hacker

idgrr committed:
Update README.md

format features and usage secions

idgrr committed:
Update README.md

Fix Spelling errors

idgrr committed:
Fix Code Quality

idgrr committed:
Merge branch 'master' of https://github.com/idgrr/ip

idgrr committed:
Rename Duke to TaskMan

idgrr committed:
Fix package naming

idgrr committed:
Fix shadowJar launcher settings

idgrr committed:
Update README.md

idgrr committed:
Set theme jekyll-theme-midnight

idgrr committed:
Update README.md

idgrr committed:
Update README.md

idgrr committed:
Update README.md

idgrr committed:
Update README.md

idgrr committed:
Merge branch 'master' of https://github.com/idgrr/ip

idgrr committed:
Fix Checkstyle

idgrr committed:
Fix Checkstyle

idgrr committed:
Fix Application size

Previously too large for small monitors
